src/lib.rs

```diff
@@ -0,0 +1,976 @@
+
+    #[inline(always)]
+    pub fn pos(self) -> OperandPos {
+        let pos_field = (self.bits >> 26) & 1;
+        match pos_field {
+            0 => OperandPos::Before,
+            1 => OperandPos::After,
+            _ => unreachable!(),
+        }
+    }
+
+    #[inline(always)]
+    pub fn constraint(self) -> OperandConstraint {
+        let constraint_field = (self.bits >> 29) & 7;
+        let preg_field = ((self.bits >> 20) as usize) & PReg::MAX;
+        match constraint_field {
+            0 => OperandConstraint::Any,
+            1 => OperandConstraint::Reg,
+            2 => OperandConstraint::Stack,
+            3 => OperandConstraint::FixedReg(PReg::new(preg_field, self.class())),
+            4 => OperandConstraint::Reuse(preg_field),
+            _ => unreachable!(),
+        }
+    }
+
+    #[inline(always)]
+    pub fn bits(self) -> u32 {
+        self.bits
+    }
+
+    #[inline(always)]
+    pub fn from_bits(bits: u32) -> Self {
+        debug_assert!(bits >> 29 <= 4);
+        Operand { bits }
+    }
+}
+
+impl std::fmt::Debug for Operand {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        std::fmt::Display::fmt(self, f)
+    }
+}
+
+impl std::fmt::Display for Operand {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match (self.kind(), self.pos()) {
+            (OperandKind::Def, OperandPos::After)
+            | (OperandKind::Mod | OperandKind::Use, OperandPos::Before) => {
+                write!(f, "{:?}", self.kind())?;
+            }
+            _ => {
+                write!(f, "{:?}@{:?}", self.kind(), self.pos())?;
+            }
+        }
+        write!(
+            f,
+            ": {}{} {}",
+            self.vreg(),
+            match self.class() {
+                RegClass::Int => "i",
+                RegClass::Float => "f",
+            },
+            self.constraint()
+        )
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum OperandConstraint {
+    /// Any location is fine (register or stack slot).
+    Any,
+    /// Operand must be in a register. Register is read-only for Uses.
+    Reg,
+    /// Operand must be on the stack.
+    Stack,
+    /// Operand must be in a fixed register.
+    FixedReg(PReg),
+    /// On defs only: reuse a use's register. Which use is given by `preg` field.
+    Reuse(usize),
+}
+
+impl std::fmt::Display for OperandConstraint {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            Self::Any => write!(f, "any"),
+            Self::Reg => write!(f, "reg"),
+            Self::Stack => write!(f, "stack"),
+            Self::FixedReg(preg) => write!(f, "fixed({})", preg),
+            Self::Reuse(idx) => write!(f, "reuse({})", idx),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum OperandKind {
+    Def = 0,
+    Mod = 1,
+    Use = 2,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum OperandPos {
```

Comment:
Yefah, I've gone back and forth on this, but on balance at least I find "before" and "after" clearer than "use" and "def". The latter are really a way of interacting with a value (consuming it or producing it), and their association with locations in the program is purely a convention... "before" and "after" on the other hand at least tell me which one is which more intuitively, though I do get the "separate from the inst, vs. attached to the edge of it" difference in connotations. I can't think of something both better and suitably pithy right now, so I'll leave this as-is, but if someone has a suggestion that's both, I'm open to it!

Reply:
Ionmonkey has used "start" and "end" which looks simple, but has triggered very large comments explaining the differences in usage (of functions like `useAtStart`, etc.). Considering that there is no obvious perfect solution, and that the literature seems to use `use` points / `def` points largely (and not "before"/"after"), and it's nicely short (can't beat 3 letters long!), I would slightly vote in favor of those, but only with my weak outsider observer voting power :)

Reply:
I've seen `Def`/`EarlyDef`/`Use`/`LateUse` used in some compilers, perhaps that works as an alternative naming scheme?

Reply:
Ah, yes, I like the "early def" / "late use" terminology; so how about `Early` / `Late`? That gets the "actually part of the instruction, not before/after it" semantics that @julian-seward1 was concerned about, and also disambiguates these from the actual `Before`/`After` insertion points for edits. I'd still prefer to avoid simple `use`/`def` because, to me, it's a category error (kind of use, not location of use) and leads to confusing things like "def at use-point" but the separate Early/Late terms with no other connotations seem less likely to cause unwanted aliasing with other meanings...